### ISSUE TITLE
PostMeta and TermMeta

### ISSUE THREAD

**Issue by lloc** [simonwheatley/wp-post-term-locale#3]
One could argue that we have to use the PostMeta and the _new_ [TermMeta](https://core.trac.wordpress.org/ticket/10142) to store additional information like the locale of the contents. I think that we should measure the performance of such a solution too to have at least an idea about the impact.

**Comment by tfrommen** [comment]
In my opinion, this is the worst approach (of the three). :)

Why is that?  
- If objects (e.g., posts, terms) are decided to have a locale, then this locale should actually be part of the object (definition), and not metadata.
- When using the meta tables, we cannot (easily) add additional information to a locale (e.g., different ISO codes, custom names whatever), which we could in case of a separate `wp_locales` table, if we wanted.
- Changing a locale would require changing the according metadata entry of every single object that has this locale, compared to changing only a single entry (the locale itself).

**Comment by lloc** [comment]
I absolutely agree but I think it would be helpful to have some metrics when want to convince other devs that ours is the direction to go.

**Comment by fgirardey** [comment]
In my opinion the first point of @tfrommen is enough to cut the deal with post meta and metrics are not involved 😊

**Comment by lloc** [comment]
As I already said I agree but I believe - because we have to convince other developers that are not that much interested - it wants a little bit more. ;) I think that point 2 

> When using the meta tables, we cannot (easily) add additional information 
> to a locale (e.g., different ISO codes, custom names whatever), which we could in case of a separate
> wp_locales table, if we wanted.

is the real win. If not we might end up with an additional `varchar` column in `wp_posts` where we store just the language codes. ;)

**Comment by fgirardey** [comment]
Good point for other developers :+1:

**Comment by markhowellsmead** [comment]
`wp_locales` sounds like a good idea; there are standards, after all, in locale definitions. The most accurate is a five character mix of ISO 639-1 for the language and ISO 3166-1 for the country code. For example, de_DE, de_CH. Then, various other codes and references can be stored in `wp_locales`: for example, the individual 2-digit codes.

**Comment by lloc** [comment]
But don't forget codes like `de_DE-formal` ;)

**Comment by markhowellsmead** [comment]
@lloc I know what you mean, but `de_DE-formal` is WordPress-specific. The development should adhere to standard language/locale pairs as far as possible, to maintain standard usage.

I doubt that there is the need to be able to switch a website content between formal and informal, like you would between different languages. For the same reason, it's unlikely that formal and informal translation files would be in use in the same installation.

**Comment by lloc** [comment]
@mhmli It should work also for multisites where this can occur easily. In my experience users will ask for it. ;)

**Comment by markhowellsmead** [comment]
Really? There are websites where website visitors can switch between a formal version of the content and an informal version? Do you have a link as an example?

**Comment by tfrommen** [comment]
Re @mhmli:

> I know what you mean, but `de_DE-formal` is WordPress-specific.

[No, it's not](https://www.gnu.org/software/gettext/manual/html_node/Locale-Names.html). ;) At least not **having** formal language variants. How they are named is another thing, yes. See also [this ticket](https://core.trac.wordpress.org/ticket/28303) for related discussion (even though it's about updating language packs).

**Comment by lloc** [comment]
@mhmli Why should we limit all this to the website visitors? ;) Just imagine a business model like SaaS that runs on a multisite.

**Comment by markhowellsmead** [comment]
@tfrommen I stand corrected :)

@lloc I don't understand your answer, sorry. The point I wanted to make is that a piece of content is always written with a specific language/locale combination in mind. If there is an installation with, for example, two German versions, then if the secondary piece of content is identical to the primary one, then this can be handled by the theme or plugin. Assigning one database record to two completely different locale/language combinations is messy.

**Comment by lloc** [comment]
@mhmli OK I'll give you an example of what I see every day. In my company we create content for our customers and push that to their websites. Our customers may generate marketing campaigns for different age-groups so they probably create separate landing pages and speaking to a visitor in a formal or informal way could make the difference.
